design/README.md

```diff
@@ -0,0 +1,169 @@
+- DIS observables (APFEL),
+- Fixed-target DY (APFEL)
+- APPLgrids for hadronic processes (APPLgrid),
+- PineAPPL grids for processes with EW corrections (PineAPPL).
+
+### Problems to be avoided
+
+Here a short summary of practical problems we would like to avoid:
+
+- avoid different data input format, today we have to maintain parsers for APFEL
+  (DIS/FTDY), APPLgrid and PineAPPL,
+- avoid long runtime operations _Proposed Solution_: use caching and store the
+  elements in a server a priori, e.g.: DGLAP evolution kernels, DIS and FTDY
+  grids,
+- reduce to minimum data input processing,
+
+## Proposal
+
+We propose to develop a code which performs the combination and provide tools to
+prepare, store and update its requirements.
+
+### Desired Features
+
+- delegate the FK Table construction to a single code/function,
+- provide tools to cache results and upload to server
+- provide tools to optimize the quality of FK tables
+
+### Input
+
+The program should take as input:
+
+1. the theory number identification number, using the NNPDF theory database.
+   Note that in opposite to apfelcomb we should take the number and not the full
+   sqlite string.
+2. the dataset name following the NNPDF convention.
+
+### Procedure
+
+1. Retrieve the EKO
+   1. Input: generate runcards
+   2. depends on 1.1: Download: we compute the md5sum or whatever other
+      algorithm on the runcard file and check for a corresponding file in the
+      server, if it doesn't exit.
+   3. depends on 1.1, fires if 1.2 fails: Create: invoke eko with the
+      configuration from (1.1) and let it compute the operator
+   4. depends on 1.3: Upload: uploads the operator from (1.3) (takes the md5sum
+      of the runcard)
+2. Retrieve PineAPPL grid
+   1. **TBD** define an analog of 1.
+3. Operator Application
+   1. do the matrix multiplication
+4. Delivery
+   1. store the generated FKTables in the NNPDF repository
+5. Quality Control
+   1. the PineAPPL predictions
+   2. the predictions using EKO with a reduced number of grid points (measures
+      the compression penalty)
+   3. the prediction of the FK table (measures the effect of the down-evolution)
+
+### Requirements for external programs
+
+#### eko
+
+- the output MUST be provided in a self-contained format, i.e. the python
+  library `eko` SHOULD NOT be required
+- the current program design fulfills this specification
+
+The current proposal for the `eko` output format is the following:
+
+- for each pair of `(Q2_0, Q2)` there will be a rank 4 tensor `(fout, xout, fin, xin)`, saved in _numpy_
+  binary format
+  - it is possible to include binary in a `yaml` or to make an explicit
+    `npy/npz` file instead
+- since `numpy` it's loosing the information on the `dtype` and `shape` on save
+  it should be shipped together in a metadata file
+- since `eko` it's able to produce operators for multiple `Q2` at the same time
+  it is possible to collect them in a single compressed archive
```

Comment:
This point is delicate:
- for `eko` purpose shipping a single zipped archive for all the generated operators it's much more natural
- this will couple arbitrarily objects that are computed together, only because of creation time, and may break (or make it much more difficult) the `md5sum` implementation

Reply:
Agreed. Unless we need the grouping for storage size or something it's better to have them all separated.